quinn/src/runtime.rs

```diff
@@ -66,6 +71,53 @@ pub trait AsyncUdpSocket: Send + Sync + Debug + 'static {
     }
 }
 
+/// Helper to coordinate concurrently writing to an [`AsyncUdpSocket`]
+pub trait UdpPoller: Send + Sync + Debug + 'static {
+    /// Check whether the associated socket is likely to be writable
+    ///
+    /// Must be called after [`AsyncUdpSocket::try_send`] returns [`io::ErrorKind::WouldBlock`] to
+    /// register the task associated with `cx` to be woken when a send should be attempted again.
+    fn poll_writable(self: Pin<&mut Self>, cx: &mut Context) -> Poll<io::Result<()>>;
+}
+
+pin_project_lite::pin_project! {
+    struct UdpPollHelper<F, T> {
+        f: F,
+        #[pin]
+        fut: Option<T>,
+    }
+}
+
+impl<F, T> UdpPollHelper<F, T> {
+    fn new(f: F) -> Self {
+        Self { f, fut: None }
+    }
+}
+
+impl<F, T> UdpPoller for UdpPollHelper<F, T>
+where
+    F: Fn() -> T + Send + Sync + 'static,
+    T: Future<Output = io::Result<()>> + Send + Sync + 'static,
+{
+    fn poll_writable(self: Pin<&mut Self>, cx: &mut Context) -> Poll<io::Result<()>> {
+        let mut this = self.project();
+        if this.fut.is_none() {
+            this.fut.set(Some((this.f)()));
+        }
+        let result = this.fut.as_mut().as_pin_mut().unwrap().poll(cx);
```

Comment:
I think you could do the option dance here to avoid using `.unwrap()`: take out the future or create it, poll it, put it back (or not if consumed).

Reply:
I think that would be unsound, because `T` may be `!Unpin`.